spec/lib/twingly/url_spec.rb

```diff
@@ -0,0 +1,455 @@
+    end
+  end
+
+  describe "#initialize" do
+    context "when given input parameters of wrong types" do
+      it "raises an error" do
+        expect { described_class.new("a", "b") }.to raise_error
+      end
+    end
+  end
+
+  describe "#scheme" do
+    subject { url.scheme }
+    it { is_expected.to eq("http") }
+  end
+
+  describe "#trd" do
+    subject { url.trd }
+    it { is_expected.to eq("www.blog") }
+  end
+
+  describe "#sld" do
+    subject { url.sld }
+    it { is_expected.to eq("twingly") }
+  end
+
+  describe "#tld" do
+    subject { url.tld }
+    it { is_expected.to eq("co.uk") }
+  end
+
+  describe "#domain" do
+    subject { url.domain }
+    it { is_expected.to eq("twingly.co.uk") }
+  end
+
+  describe "#host" do
+    subject { url.host }
+    it { is_expected.to eq("www.blog.twingly.co.uk") }
+  end
+
+  describe "#origin" do
+    subject { url.origin }
+    it { is_expected.to eq("http://www.blog.twingly.co.uk") }
+  end
+
+  describe "#path" do
+    subject { url.path }
+    it { is_expected.to eq("/2015/07/01/language-detection-changes/") }
+  end
+
+  describe "#normalized_path" do
+    subject { url.normalized_path }
+    it { is_expected.to eq("/2015/07/01/language-detection-changes") }
+  end
+
+  describe "#normalized_scheme" do
+    subject { url.normalized_scheme }
+    it { is_expected.to eq("http") }
+  end
+
+  describe "#normalized_host" do
+    subject { url.normalized_host }
+    it { is_expected.to eq("www.blog.twingly.co.uk") }
+  end
+
+  describe "#valid?" do
+    invalid_urls.each do |invalid_url|
+      it "returns false for an invalid URL (#{invalid_url})" do
+        expect(described_class.parse(invalid_url).valid?).to be false
+      end
+    end
+
+    %w(http://blog.twingly.com/ hTTP://blog.twingly.com/ https://blog.twingly.com).each do |valid_url|
+      it "returns true for the valid url '#{valid_url}'" do
+        expect(described_class.parse(valid_url).valid?).to be true
+      end
+    end
+
+    context "when given nil input" do
+      it "it returns false" do
+        expect(described_class.parse(nil).valid?).to be false
+      end
+    end
+  end
+
+  describe "#normalized" do
+    subject { described_class.parse(url).normalized.to_s }
+
+    context "adds www if host is missing a subdomain" do
+      let(:url)      { "http://twingly.com/" }
+      let(:expected) { "http://www.twingly.com/" }
+
+      it { is_expected.to eq(expected) }
+    end
+
+    context "does not add www if the host has a subdomain" do
+      let(:url) { "http://blog.twingly.com/" }
+
+      it { is_expected.to eq(url) }
+    end
+
+    context "does not remove www if the host has a subdomain" do
+      let(:url) { "http://www.blog.twingly.com/" }
+
+      it { is_expected.to eq(url) }
+    end
+
+    context "keeps www if the host already has it" do
+      let(:url) { "http://www.twingly.com/" }
+
+      it { is_expected.to eq(url) }
+    end
+
+    context "adds a trailing slash if missing in origin" do
```

Comment:
`/` is not part of origin (which is scheme, host and port), so the wording is wrong here